### ISSUE TITLE
dark mode

### ISSUE THREAD

**Issue by shriram** [brownplt/code.pyret.org#329]
Is there any chance of getting a dark mode (or even just switching to dark)? This seems to be by far the most popular option with Web IDEs these days (and frankly it is pretty pleasant on the eyes). Similarly, there seems to be a trend towards using a larger font size by default, and towards trying to use more pleasing fonts. (These would also help avoid one/ell confusion.) The CPO look feels a bit dated in contrast.

See, for instance, codepen.io; https://codepen.io/aaroniker/pen/pojXjrZ is a useful representative that illustrates the background, font-size, and font. They seem to use 15pt Monaco (with MonoSpace as the backup). Monaco is very pleasant, and does clearly differentiate characters.

From what I can tell, about a dozen options in src/web/css/editor.css and src/web/css/reset.css control all these options. I've been able to successfully change almost all these options to get a quasi-working "dark mode", but I know a lot of thought went into Ben's selections (especially where he made HSL choices); e.g., I haven't been able to get a satisfactory look for the span, which comes out ghosty.

But even a change of font and size would be a great start, I think.

**Comment by blerner** [comment]
Dark themes are contentious; I happen to strongly dislike the codepen theme, actually :-\  (Something about the olive/khaki colors makes me slightly nauseated.)

We spent a lot of time working through the colors for CPO so that they'd match _whiteboard_ experiences for teachers.  If we really want to spend time thinking through what a darkboard experience should be (I can't say "blackboard", since that implies chalk colors!), make some concrete proposals and we can start workshopping it.

As for the font -- on Windows for me, I get 14px Courier New, not 15pt Monaco.  (I'm not at my Linux machine right now to see what it gets instead.)  CPO gives me 12px Courier New.  I wouldn't mind 14px as a new default, though I think 16px (which is the next size up in the Bonnie menu) is too big as a default, and too small anyway for teachers' presentation mode.  We can add some other fonts as defaults before the monospace fallback that we currently have.  But we'd need to pick a font that most of our students are likely to already have on their machines, or else it's not gonna make a difference.

**Comment by shriram** [comment]
There was a reason I suggested their font and size but not specific colors (-:.

TBH I'm not sure how much the whiteboard matching matters any longer. Most of our users are high school teachers who I think work with our materials electronically; that's about to become even more so now that teachers are going to be getting slide decks with embedded CPOs.

I agree that 16pt is way too big to start with. I'm fine with 14. (I believe there are some interline spacing constraints in the CSS that are hard-coded, so they will need updating too.)

As for fonts, I really like Consolas; that's what I use for my slides. That may well be available on both platforms, since PowerPoint offers it? I'll also ask on SM.

Let's at least do the font and size and see how that "feels".

**Comment by sorawee** [comment]
Well, it exists already as an extension: https://github.com/thomascastleman/dark-pyret / https://chrome.google.com/webstore/detail/dark-pyret/jahmdeebfkoocpblokbimjnebdogpflo

**Comment by shriram** [comment]
WAT.

**Comment by shriram** [comment]
Most popular Twitter reco is for FiraCode: https://fonts.google.com/specimen/Fira+Code. It's served by Google (so it should be just as unblocked as Google Classroom, etc., in schools). And it has really nice ligatures for many common programming constructs. People like how things vertically align.

Next on the list are Losevka (not on Google Fonts), Source Code [https://fonts.google.com/specimen/Source+Code+Pro?query=source+code], and Hack (not…).

Overall vote summary (ignoring Nate Foster's "Comic Sans"):

Fira[Code] 9
Losevka 7
Source Code 7
DejaVu 5
Hack 5
IBM Plex 4
Inconsolata 4
Menlo 4
Pragmata 3
Cascadia 2
Fantasque 2
JetBrains 2
Roboto 2
DOSEGA
Dank
GNU FreeFont
Go
Libertine
Monoid
Mononoki
Oxygen
Ubuntu Monospace
Victor

**Comment by jpolitz** [comment]
I don't recommend Fira Code for learning if we have the ligatures on (though I like it for myself).

I've livecoded with it and undergrads get confused about the ligatures – if they miss the -> turning into → they wonder “where's the arrow key.” A reasonable confusion!

If we use it in a mode without ligatures, I think it's a great font.

**Comment by ds26gte** [comment]
Consolas truly is a great coding font as Shriram said (I've used only it
forever), but since it isn't a Web font, perhaps its closest Google-font
equivalent, Inconsolata (the name seems to be a deliberate homage), is a
good substitute?

--d

**Comment by ds26gte** [comment]
P.S. Never mind, Inconsolata is already in Shriram's list...

**Comment by shriram** [comment]
Yes, I believe Inconsolata is a deliberate homage.

Fwiw, I’ve been spending the past few days in other IDEs. I’ve spent all of
this evening in CPO using the Dark Pyret extension [
https://github.com/thomascastleman/dark-pyret] and the font really feels
much more awkward.

**Comment by shriram** [comment]
ISWYM Joe. I went into Inspect and changed the font to Fira Code. On the
one hand, the ligatures are awesome. On the other hand, I totally see what
you mean: especially imagine an isolated teacher without anyone to turn to
quickly, stuck having to figure out and then explain all this. So at the
very least, disable the ligatures…

**Comment by blerner** [comment]
The non-ligature version of Fira Code is called Fira Mono

**Comment by jswrenn** [comment]
Ligatures can be [controlled via CSS](https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant-ligatures), so we could potentially load Fira Code and merely disable its ligatures by default.

**Comment by blerner** [comment]
There's really no point: Fira Code was a fancied-up version of Fira Mono; if we just want the baseline monospace font, then just use it directly.

**Comment by jswrenn** [comment]
Yeah; I'm imagining ligatures being an option some users would want to
enable.

**Comment by shriram** [comment]
The odds of this creating confusion (in screen-shots, pair-programming,
etc.) is just too high. We're creating a language for beginners, so let's
keep it simple. Those who really want something more sophisticated can go
in and edit font properties.

**Comment by blerner** [comment]
Ok, let's keep this discussion streamlined.  Adding ligatures to CPO is not a desired feature _for our intended audience_, so let's keep this issue mostly about dark mode.

In https://github.com/brownplt/code.pyret.org/tree/style-updates I've started work on this: I've updated the font to Fira Mono, I've updated the default font size to 14px (note: pixels, not points, because the web sucks), and I've started merging in the styles from the Dark Pyret extension.  (I've gotten in touch with the authors, and they're on board with merging their work into CPO proper.)  The styles don't quite all work correctly yet, and I'm sure they're incomplete, so take that branch as a starting point, not a ready-for-review PR.
